lightning/src/util/persist.rs

```diff
@@ -249,10 +259,1010 @@ where
+			)
+			.unwrap(),
+			index: u16::MAX,
+		});
+		assert_eq!(
+			monitor_name2.as_str(),
+			"f33dbeeff33dbeeff33dbeeff33dbeeff33dbeeff33dbeeff33dbeeff33dbeef_65535"
+		);
+	}
+
+	#[test]
+	fn bad_monitor_string_fails() {
+		assert!(MonitorName::new(
+			"deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef".to_string()
+		)
+		.is_err());
+		assert!(MonitorName::new(
+			"deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef_65536".to_string()
+		)
+		.is_err());
+		assert!(MonitorName::new(
+			"deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef_21".to_string()
+		)
+		.is_err());
+	}
+
+	#[test]
+	fn persister_with_real_monitors() {
+		let test_max_pending_updates = 7;
+		let chanmon_cfgs = create_chanmon_cfgs(4);
+		let persister_0 = MonitorUpdatingPersister {
+			kv_store: &TestStore::new(false),
+			logger: &TestLogger::new(),
+			maximum_pending_updates: test_max_pending_updates,
+			entropy_source: &chanmon_cfgs[0].keys_manager,
+			signer_provider: &chanmon_cfgs[0].keys_manager,
+		};
+		let persister_1 = MonitorUpdatingPersister {
+			kv_store: &TestStore::new(false),
+			logger: &TestLogger::new(),
+			maximum_pending_updates: test_max_pending_updates,
+			entropy_source: &chanmon_cfgs[1].keys_manager,
+			signer_provider: &chanmon_cfgs[1].keys_manager,
+		};
+		let mut node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+		let chain_mon_0 = test_utils::TestChainMonitor::new(
+			Some(&chanmon_cfgs[0].chain_source),
+			&chanmon_cfgs[0].tx_broadcaster,
+			&chanmon_cfgs[0].logger,
+			&chanmon_cfgs[0].fee_estimator,
+			&persister_0,
+			&chanmon_cfgs[0].keys_manager,
+		);
+		let chain_mon_1 = test_utils::TestChainMonitor::new(
+			Some(&chanmon_cfgs[1].chain_source),
+			&chanmon_cfgs[1].tx_broadcaster,
+			&chanmon_cfgs[1].logger,
+			&chanmon_cfgs[1].fee_estimator,
+			&persister_1,
+			&chanmon_cfgs[1].keys_manager,
+		);
+		node_cfgs[0].chain_monitor = chain_mon_0;
+		node_cfgs[1].chain_monitor = chain_mon_1;
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+		let broadcaster_0 = &chanmon_cfgs[2].tx_broadcaster;
+		let broadcaster_1 = &chanmon_cfgs[3].tx_broadcaster;
+
+		// Check that the persisted channel data is empty before any channels are
+		// open.
+		let mut persisted_chan_data_0 = persister_0
+			.read_all_channel_monitors_with_updates(&broadcaster_0, &chanmon_cfgs[0].fee_estimator)
+			.unwrap();
+		assert_eq!(persisted_chan_data_0.len(), 0);
+		let mut persisted_chan_data_1 = persister_1
+			.read_all_channel_monitors_with_updates(&broadcaster_1, &chanmon_cfgs[1].fee_estimator)
+			.unwrap();
+		assert_eq!(persisted_chan_data_1.len(), 0);
+
+		// Helper to make sure the channel is on the expected update ID.
+		macro_rules! check_persisted_data {
+			($expected_update_id: expr) => {
+				persisted_chan_data_0 = persister_0
+					.read_all_channel_monitors_with_updates(
```

Comment:
That would make the assertion that we only stored one channel monitor always true. Seems like a step back. It would also mean I'd need another argument in this macro, since I'd need to know the monitor key. We're listing out of memory here so I don't think speed is worth much.